### PR TITLE
update ldap configuration

### PR DESCRIPTION
**Pull request by wemeya**
an example of we discussed, some still need to prove.

### CODE REVIEW

File: auth/ldap/ldap.go [hunk 1]
```diff
@@ -65,16 +65,65 @@ func (l *Auth) Authenticate(m models.AuthModel) (*models.User, error) {
...
+	}else {
+		scope = openldap.LDAP_SCOPE_SUBTREE
+	}
+	attributes := []string{"cn", "mail", "email", "uid"}
+	result, err := ldap.SearchAll(ldapBaseDn, scope, filter, attributes)
+	if err != nil {
+		return nil, err
+	}
+	if len(result.Entries()) != 1 {
+		log.Warningf("Found more than one entry.")
```
Comment (reasonerjt) [inline auth/ldap/ldap.go]: Could you dump more debug information such as base dn and filters?

File: auth/ldap/ldap.go [hunk 2]
```diff
@@ -65,16 +65,65 @@ func (l *Auth) Authenticate(m models.AuthModel) (*models.User, error) {
...
+			u.Email = val
+		//some use email
+		case "email":
+			u.Email = val
+		//AD use cn
+		case "cn":
+			u.Username = val
+		}
+	}	
+	log.Debug("Let me see bindDN (result.Base): ", result.Base())
```
Comment (reasonerjt) [inline auth/ldap/ldap.go]: What's this?

File: auth/ldap/ldap.go [hunk 3]
```diff
@@ -65,16 +65,65 @@ func (l *Auth) Authenticate(m models.AuthModel) (*models.User, error) {
...
 
-	err = ldap.Bind(baseDn, m.Password)
+	err = ldap.Bind(bindDN, m.Password)
```
Comment (reasonerjt) [inline auth/ldap/ldap.go]: maybe you should first check the password and if it passed, populate the attributes?

File: Deploy/harbor.cfg [hunk 4]
```diff
@@ -2,7 +2,8 @@
 
 #The IP address or hostname to access admin UI and registry service.
 #DO NOT use localhost or 127.0.0.1, because Harbor needs to be accessed by external clients.
-hostname = reg.mydomain.com
+
+hostname = 10.117.4.245
```
Comment (reasonerjt) [inline Deploy/harbor.cfg]: Use default values in harbor.cfg!

File: Deploy/prepare [hunk 5]
```diff
@@ -99,6 +102,9 @@ render(os.path.join(templates_dir, "ui", "env"),
         harbor_admin_password=harbor_admin_password,
         ldap_url=ldap_url,
         ldap_basedn=ldap_basedn,
+        ldap_filter=ldap_filter,
+        ldap_uid=ldap_uid,
+        ldap_scope=ldap_scope,
```
Comment (reasonerjt) [inline Deploy/prepare]: you should validate the scope? or mention any invalid values is subtree scope?

File: auth/ldap/ldap.go [hunk 6]
```diff
@@ -65,16 +65,65 @@ func (l *Auth) Authenticate(m models.AuthModel) (*models.User, error) {
...
 	if err != nil {
 		return nil, err
 	}
 	defer ldap.Close()
 
-	scope := openldap.LDAP_SCOPE_SUBTREE // LDAP_SCOPE_BASE, LDAP_SCOPE_ONELEVEL, LDAP_SCOPE_SUBTREE
+/*	scope := openldap.LDAP_SCOPE_SUBTREE // LDAP_SCOPE_BASE, LDAP_SCOPE_ONELEVEL, LDAP_SCOPE_SUBTREE
```
Comment (reasonerjt) [inline auth/ldap/ldap.go]: delete the code if not needed.

File: auth/ldap/ldap.go [hunk 7]
```diff
@@ -65,16 +65,65 @@ func (l *Auth) Authenticate(m models.AuthModel) (*models.User, error) {
...
+		return nil, err
+	}
+	if len(result.Entries()) != 1 {
+		log.Warningf("Found more than one entry.")
+		return nil, nil
+	}
+	en := result.Entries()[0]
+	bindDN := en.Dn()
+	u := models.User{}
+	for _, attr := range en.Attributes() {
```
Comment (reasonerjt) [inline auth/ldap/ldap.go]: So is it possible the entry has no uid nor cn? we probably should check that and fail before inserting to DB?  or shall we use the uid attribute user set in the .cfg as username?

**Comment by reasonerjt** [comment]
please squash the commits.

**Comment by wemeya** [comment]
@reasonerjt squashed.

**Comment by reasonerjt** [comment]
Sorry, this has to goto dev branch.... I didn't check it before

**Comment by wemeya** [comment]
ok, I will open a new PR.

**Comment by reasonerjt** [comment]
And I don't quite like the idea about searchdn, since you cannot ask user to comment something.
I'd suggest you create pr without searchdn support.

**Comment by wemeya** [comment]
So I initially comment the searchdn in harbor.cfg and if needed, ask the user themselves to uncomment it. Is that you mean?

**Pull request closed without merging**